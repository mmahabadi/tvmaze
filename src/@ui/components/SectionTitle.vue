<script setup lang="ts">
defineProps<{ title: string }>()
</script>

<template>
  <div class="my-9">
    <div class="w-full flex sm:gap-8 gap-4 items-center">
      <div class="text-teal-500"><slot></slot></div>
      <h2 class="sm:text-xl font-bold text-lg">{{ title }}</h2>
    </div>
  </div>
</template>
