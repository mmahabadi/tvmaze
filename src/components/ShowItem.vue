<script setup lang="ts">
import { RouterLink } from 'vue-router'
import { type Movie } from '@/types'
import IconPlay from '@/@ui/components/icons/IconPlay.vue'
import Image from '@/@ui/components/Image.vue'

type propType = Pick<Movie, 'name' | 'id'> & {
  image?: string
}

defineProps<propType>()
</script>
<template>
  <div class="show-item">
    <RouterLink :to="`/details/${id}`">
      <div>
        <Image :src="image" :alt="name" />
      </div>
      <div class="caption">
        <h3>{{ name }}</h3>
        <button>
          <IconPlay />
        </button>
      </div>
    </RouterLink>
  </div>
</template>

<style scoped>
.show-item {
  @apply overflow-hidden border border-gray-600 rounded-md p-1 transform duration-150;
  background-color: var(--bg-color);

  .caption {
    @apply absolute bottom-0 left-0 right-0 py-3 px-4 bg-black bg-opacity-70 text-white flex justify-between items-center;

    h3 {
      @apply overflow-hidden truncate font-semibold text-nowrap;
    }

    button {
      @apply flex justify-center items-center text-xs text-white transform duration-150 bg-white bg-opacity-20 rounded-md w-8 h-8;
      border: 2px solid var(--primary-color);
    }
  }

  &:hover {
    @apply relative scale-110 z-10;

    button {
      background-color: var(--primary-color);
    }
  }
  a {
    @apply w-full;
  }

  img {
    @apply w-full h-64 object-cover;
  }
}
</style>
