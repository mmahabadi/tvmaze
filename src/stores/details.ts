import { defineStore } from 'pinia'
import { ref } from 'vue'
import * as ApiService from '@/services/ApiService'
import type { Movie } from '@/types'

export const useDetailsStore = defineStore('details', () => {
  const details = ref<Movie | null>(null)
  const loading = ref(false)
  const error = ref<null | Error>(null)

  async function getShowDetails(id: number): Promise<void> {
    loading.value = true
    try {
      details.value = await ApiService.getDetails(id)
    } catch (e) {
      error.value = e as Error
    } finally {
      loading.value = false
    }
  }

  return {
    details,
    loading,
    error,
    getShowDetails
  }
})
