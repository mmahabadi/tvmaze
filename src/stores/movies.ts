import { defineStore } from 'pinia'
import { computed, ref, type ComputedRef } from 'vue'
import * as ApiService from '@/services/ApiService'
import { type Movie, type ShowByGenre } from '@/types'

export const useMoviesStore = defineStore('movies', () => {
  const shows = ref<Map<string, ShowByGenre>>(new Map())
  const movies = ref<Movie[]>([])
  const details = ref<Movie | null>(null)
  const loading = ref(false)
  const error = ref<null | Error>(null)
  const currentIndex = ref(1)
  const ITEMS_PER_PAGE = 3
  const ITEMS_LIMIT_FOR_RECENT_MOVIES = 15
  const DATE_LIMIT_FOR_RECENT_MOVIES = 2020

  const recentMovies = computed(() =>
    movies.value
      .filter((movie) => new Date(movie.ended).getFullYear() > DATE_LIMIT_FOR_RECENT_MOVIES)
      .slice(0, ITEMS_LIMIT_FOR_RECENT_MOVIES)
  )

  /**
   * @note because the API does not support fetching genres,
   * I need to extract the genres from the fetched movies
   */
  const genres = computed(() => [...new Set([...movies.value.flatMap((movie) => movie.genres)])])

  const getByGenre = computed(() => (genre: string): ShowByGenre => {
    loadGenre(genre)
    return shows.value.get(genre) as ShowByGenre
  })

  const paginatedGenres: ComputedRef<string[]> = computed(() => {
    const startIndex = currentIndex.value
    const endIndex = startIndex * ITEMS_PER_PAGE
    // Concatenate the new slice with the previous values of pagedGenres
    return [...genres.value.slice(0, endIndex)]
  })

  async function fetchMovies(): Promise<void> {
    loading.value = true
    try {
      movies.value = await ApiService.getData()
    } catch (e) {
      error.value = e as Error
    } finally {
      loading.value = false
    }
  }

  async function fetchShowByGenre(genre: string): Promise<void> {
    const genreShows = shows.value.get(genre)
    if (!genreShows) return

    try {
      /**
       * @note if the APi supports fetching movies by genre,
       * comment the following line
       * and uncomment the line below
       */
      const items = movies.value
        .filter((movie) => movie.genres.includes(genre))
        .sort((a, b) => b.rating.average - a.rating.average)
      /**
       * @note the API does not support fetching movies by genre
       * uncomment the following line to fetch the movies by genre from the API
       */
      // const items = await ApiService.getShowsByGenre(genre)
      if (items) {
        genreShows.movies = items
      }
    } catch (e) {
      genreShows.error = e as Error
    } finally {
      genreShows.loading = false
    }
    shows.value.set(genre, genreShows as ShowByGenre)
  }

  function loadGenre(genre: string): void {
    if (!shows.value.has(genre)) {
      const genreShows: ShowByGenre = {
        genre,
        movies: [],
        loading: true,
        error: null
      }
      shows.value.set(genre, genreShows)
      fetchShowByGenre(genre)
    }
  }

  function nextPage(): void {
    if (currentIndex.value * ITEMS_PER_PAGE <= genres.value.length) currentIndex.value++
  }

  async function getShowDetails(id: number): Promise<void> {
    loading.value = true
    try {
      details.value = await ApiService.getDetails(id)
    } catch (e) {
      error.value = e as Error
    } finally {
      loading.value = false
    }
  }

  return {
    movies,
    genres: paginatedGenres,
    recentMovies,
    details,
    loading,
    error,
    fetchMovies,
    getByGenre,
    nextPage,
    getShowDetails
  }
})
