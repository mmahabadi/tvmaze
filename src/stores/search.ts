import type { Movie } from '@/types'
import { defineStore } from 'pinia'
import { ref } from 'vue'
import * as ApiService from '@/services/ApiService'

export const useSearchStore = defineStore('search', () => {
  const searchResults = ref<Movie[]>([])
  let searchQuery = ''
  const loading = ref(false)
  const error = ref<null | Error>(null)

  async function searchShows(query: string): Promise<void> {
    //avoid making the same request multiple times
    if (searchQuery === query) return
    searchQuery = query
    loading.value = true
    try {
      searchResults.value = await ApiService.search(query)
    } catch (e) {
      error.value = e as Error
    } finally {
      loading.value = false
    }
  }

  return {
    searchResults,
    searchQuery,
    loading,
    error,
    searchShows
  }
})
