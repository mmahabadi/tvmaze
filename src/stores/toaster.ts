import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { ToasterMessage } from '@/types'

export const useToasterStore = defineStore('toaster', () => {
  const toasters = ref<ToasterMessage[]>([])
  let idCounter = 0

  const addToaster = (message: string, type: 'warning' | 'error' | 'success') => {
    toasters.value.push({ id: idCounter++, message, type })
  }

  const removeToaster = (id: number) => {
    toasters.value = toasters.value.filter((toaster) => toaster.id !== id)
  }

  return {
    toasters,
    addToaster,
    removeToaster
  }
})
