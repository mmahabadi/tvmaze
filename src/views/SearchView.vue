<script setup lang="ts">
import Loading from '@/@ui/components/LoadingOverlay.vue'
import SectionTitle from '@/@ui/components/SectionTitle.vue'
import SearchIcon from '@/@ui/components/icons/SearchIcon.vue'
import ShowItem from '@/components/ShowItem.vue'
import MainLayout from '@/layouts/MainLayout.vue'
import { useSearchStore } from '@/stores/search'
import { ref, watchEffect } from 'vue'
import { useRoute } from 'vue-router'

const store = useSearchStore()
const route = useRoute()
const query = ref('')

watchEffect(() => {
  query.value = route.query.q?.toString() || ''
})
</script>

<template>
  <MainLayout :loading="false">
    <Loading v-if="store.loading" />
    <SectionTitle :title="`Search results for '${query}'`">
      <SearchIcon />
    </SectionTitle>
    <div class="container">
      <div class="no-result" v-if="store.searchResults?.length === 0">
        <p>No results found</p>
      </div>
      <div class="item" v-for="item in store.searchResults" :key="item.id">
        <ShowItem v-if="item" :id="item.id" :name="item.name" :image="item.image?.medium" />
      </div>
    </div>
  </MainLayout>
</template>
<style scoped>
.container {
  @apply flex gap-4 flex-wrap justify-center;

  .item {
    @apply w-full md:w-1/2 lg:w-1/3 xl:w-1/4;
  }
  .no-result {
    @apply w-full  h-full flex items-center justify-center;
  }
}
</style>
