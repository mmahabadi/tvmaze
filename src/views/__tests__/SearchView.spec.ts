import { describe, it, expect, beforeEach, vi, type MockInstance } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import SearchView from '../SearchView.vue'
import { createPinia, setActivePinia } from 'pinia'
import { useRoute } from 'vue-router'
import { useSearchStore } from '@/stores/search'
import SectionTitle from '@/@ui/components/SectionTitle.vue'
import SearchIcon from '@/@ui/components/icons/SearchIcon.vue'
import ShowItem from '@/components/ShowItem.vue'

vi.mock('vue-router', () => ({
  useRoute: vi.fn()
}))

const mockRoute = {
  query: {
    q: 'test'
  }
}

describe('SearchView', () => {
  let wrapper: VueWrapper<unknown>
  let fetchShowsSpy: MockInstance<[q: string], Promise<void>>

  beforeEach(() => {
    setActivePinia(createPinia())
    ;(useRoute as any).mockReturnValue(mockRoute)

    const store = useSearchStore()
    fetchShowsSpy = vi.spyOn(store, 'searchShows')

    wrapper = mount(SearchView, {
      global: {
        stubs: {
          MainLayout: true,
          SectionTitle,
          SearchIcon,
          ShowItem
        },
        provide: {
          store
        }
      }
    })
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  // it('calls fetchShows on mount', () => {
  //   expect(fetchShowsSpy).toHaveBeenCalledWith(mockRoute.query.q)
  // })
})
