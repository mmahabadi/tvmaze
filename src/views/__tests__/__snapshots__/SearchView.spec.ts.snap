// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchView > renders correctly 1`] = `"<main-layout-stub data-v-61d5b3d3="" loading="false"></main-layout-stub>"`;
